pypesto/hierarchical/util.py

```diff
@@ -271,6 +285,149 @@ def apply_sigma(
+
+    Computes the optimal scaling and offset of a constrained optimization in
+    case the unconstrained optimization yields a value outside the bounds.
+    We know the optimal solution then lies on the boundary of the bounds.
+    If both parameters are unsatisfied, we need to check 2 unconstrained
+    problems, and evaluate at 3 edge points. Otherwise, if only one parameter
+    is unsatisfied, we need to check 1 unconstrained problem, and evaluate at
+    2 edge points. We then return the candidate point with the lowest objective
+    value.
+
+    Parameters
+    ----------
+    data:
+        The data.
+    sim:
+        The simulation.
+    sigma:
+        The noise parameters.
+    s:
+        The scaling parameter.
+    b:
+        The offset parameter.
+    s_opt_value:
+        The optimal scaling value of the unconstrained problem.
+    b_opt_value:
+        The optimal offset value of the unconstrained problem.
+
+    Returns
+    -------
+    The optimal scaling and offset of the constrained problem.
+    """
+    # Define relevant data and sim
+    # Make all non-masked data and sim nan's in the original one
+    relevant_data = copy.deepcopy(data)
+    relevant_sim = copy.deepcopy(sim)
+    for i in range(len(data)):
+        relevant_data[i][~s.ixs[i]] = np.nan
+        relevant_sim[i][~s.ixs[i]] = np.nan
+
+    # Get bounds
+    s_bounds = s.get_bounds()
+    b_bounds = b.get_bounds()
+
+    # Get unsatisfied bounds
+    s_unsatisfied = s.get_unsatisfied_bound_index(s_opt_value)
+    b_unsatisfied = b.get_unsatisfied_bound_index(b_opt_value)
+
+    # If both parameters are unsatisfied, we need to check 2
+    # unconstrained problems, and evaluate at 3 boundary points
+    if s_unsatisfied is not None and b_unsatisfied is not None:
+        candidate_points = [
+            (s_bounds[s_unsatisfied], b_bounds[b_unsatisfied]),
+            (s_bounds[s_unsatisfied], b_bounds[1 - b_unsatisfied]),
+            (s_bounds[1 - s_unsatisfied], b_bounds[b_unsatisfied]),
+        ]
+
+        # Solve the two unconstrained problems
+        candidate_optimal_solutions = [
+            (
+                s_bounds[s_unsatisfied],
+                compute_optimal_offset(
+                    data, sim, sigma, s.ixs, s_bounds[s_unsatisfied]
+                ),
+            ),
+            (
+                compute_optimal_scaling(
+                    data, sim, sigma, s.ixs, b_bounds[b_unsatisfied]
+                ),
+                b_bounds[b_unsatisfied],
+            ),
+        ]
+
+        # Check if the unconstrained solutions are in the bounds
+        # if so, add then to the candidate points
+        for candidate_optimal_solution in candidate_optimal_solutions:
+            if s.check_within_bounds(
+                candidate_optimal_solution[0]
+            ) and b.check_within_bounds(candidate_optimal_solution[1]):
+                candidate_points.append(candidate_optimal_solution)
```

Comment:
I think you can skip checking the vertices manually. Instead, do the edge optimization first, and bound the result, which will yield the correct vertex if the optimum lies outside the bounds. This means you only need to recompute the NLLH if a vertex was chosen, otherwise you can re-use the optimization result.

Reply:
True, the implementation is now so much smaller and nicer. I just need to clip any constrained optimization and return it in case of 1 unsatisfied bound. In case of 2 I evaluate the clipped solutions and return the better one.